src/core/border_router/routing_manager.cpp

```diff
@@ -373,22 +471,48 @@ bool RoutingManager::ContainsPrefix(const Ip6::Prefix &aPrefix, const Ip6::Prefi
     return ret;
 }
 
-const Ip6::Prefix *RoutingManager::EvaluateOnLinkPrefix(void) const
+const Ip6::Prefix *RoutingManager::EvaluateOnLinkPrefix(void)
 {
-    const Ip6::Prefix *newOnLinkPrefix = nullptr;
+    const Ip6::Prefix *newOnLinkPrefix      = nullptr;
+    const Ip6::Prefix *smallestOnLinkPrefix = nullptr;
+    uint8_t            onLinkPrefixNum      = 0;
 
-    // We don't evaluate on-link prefix if we are doing
-    // Router Discovery or we have already discovered some
-    // on-link prefixes.
+    // We don't evaluate on-link prefix if we are doing Router Solicitation.
     VerifyOrExit(!mRouterSolicitTimer.IsRunning());
-    if (IsValidOnLinkPrefix(mDiscoveredOnLinkPrefix))
+
+    for (uint8_t i = 0; i < mDiscoveredPrefixNum; ++i)
     {
-        otLogInfoBr("EvaluateOnLinkPrefix: there is already on-link prefix %s on interface %u",
-                    mDiscoveredOnLinkPrefix.ToString().AsCString(), mInfraIfIndex);
-        ExitNow();
+        if (!mDiscoveredPrefixes[i].mIsOnLinkPrefix)
+        {
+            continue;
+        }
+
+        ++onLinkPrefixNum;
+
+        if (smallestOnLinkPrefix == nullptr ||
+            IsPrefixSmallerThan(mDiscoveredPrefixes[i].mPrefix, *smallestOnLinkPrefix))
+        {
+            smallestOnLinkPrefix = &mDiscoveredPrefixes[i].mPrefix;
+        }
     }
 
-    newOnLinkPrefix = &mLocalOnLinkPrefix;
+    // We start advertising our local on-link prefix if there is no existing one.
+    if (onLinkPrefixNum == 0)
+    {
+        if (AddExternalRoute(mLocalOnLinkPrefix, OT_ROUTE_PREFERENCE_MED) == OT_ERROR_NONE)
+        {
+            newOnLinkPrefix = &mLocalOnLinkPrefix;
+        }
+    }
+    // We may not be able to remove our local on-link prefix when a bigger prefix is received.
```

Comment:
In case an application-specific on-link prefix is received and it is bigger than our advertised prefix, both prefixes will stay on the infra link. Updated the documentation, please review.